### ISSUE TITLE
Debug flag support for different Intel compiler versions

### ISSUE THREAD

**Issue by lizziel** [GEOS-ESM/ESMA_cmake#125]
ESMA_cmake debug flags used with Intel compilers include unitialized variables check `-check uninit`. 

https://github.com/GEOS-ESM/ESMA_cmake/blob/c2e6fc5c9903624a443f96b3a50c62f22aad5a3d/Intel.cmake#L58

This works with ifort19 but causes an error with ifort18 when building MAPL, as reported by a GCHPctm user (https://github.com/geoschem/GCHPctm/issues/35). The error message is:

[ 15%] Building Fortran object src/MAPL/MAPL_cfio_r4/CMakeFiles/MAPL_cfio_r4.dir/ESMF_CFIOBaseMod.f.o
ifort: error #10236: File not found:  'uninit'
make[2]: *** [src/MAPL/MAPL_cfio_r4/CMakeFiles/MAPL_cfio_r4.dir/ESMF_CFIOBaseMod.f.o] Error 1
make[2]: *** Deleting file `src/MAPL/MAPL_cfio_r4/CMakeFiles/MAPL_cfio_r4.dir/ESMF_CFIOBaseMod.f.o'
make[1]: *** [src/MAPL/MAPL_cfio_r4/CMakeFiles/MAPL_cfio_r4.dir/all] Error 2
make: *** [all] Error 2

Changing the flag `-check,uninit` fixes the problem, but then makes it incompatible with ifort19.

Do you have plans to allow for version-specific flags in ESMA_cmake to fix issues like this?

**Comment by tclune** [comment]
Hmm.  I thought `-check unit` was supported for quite some time.

**Comment by lizziel** [comment]
Looks like I misread the original poster's suggested fix. It is actually this: `I recommend changing -check bounds -check uninit to -check bounds,uninit`

**Comment by tclune** [comment]
The error must be more subtle.   I've confirmed that on our cluster Intel 17, 18 and 19 all support `check unit`.  (Further, we routinely build GEOS using Intel 18 and the same `Intel.cmake`.

Could the user provide us with (a) the output from the cmake config step and (b) the log for a serial build using `VERBOSE=1`?    I want to see what command and flags is actually being issued to the compiler.

**Comment by tclune** [comment]
@lizziel Our messages crossed each other.    I'd still like to see what is actually going wrong.   I know there are parts of the model where @mathomp4 had to do some fairly ugly things to get compiler flags in the right order, but ... this is in MAPL.  (And again, the existing logic works for us.)

Regardless of any bugs, I defer to @mathomp4 as to whether the concatenate form for the compiler flags is easier for us to work with or not.

**Comment by lizziel** [comment]
I reproduced the issue on my end using the original flags in Intel.cmake.  I am using the following libraries:
Intel/18.0.5
OpenMPI/4.0.1
CMake/3.16.1

Here are the logs:
[configure.txt](https://github.com/GEOS-ESM/ESMA_cmake/files/5152578/configure.txt)
[compile.txt](https://github.com/GEOS-ESM/ESMA_cmake/files/5152579/compile.txt)
[compile_verbose.txt](https://github.com/GEOS-ESM/ESMA_cmake/files/5152580/compile_verbose.txt)

**Comment by tclune** [comment]
Ah the error is happening in pFlogger.  That's an important clue, as we build pFlogger separately at this end.    I'll have to dig down to see how the esma flags override the ones in pFlogger.   And it begs the question as to where the bug really is in any event.

**Comment by tclune** [comment]
@lizziel   Looking closer,  I see that your logs are insufficent.   `compile.txt` was parallel which mixed the error messages, so that I can't see which files actually had the error.  Maybe it is pFlogger, maybe it was an earlier file in the log.    And the `compile_verbose.txt` does not show any errors.  Just stops at 33%.   Maybe you hit an upload limit on GitHub?    If you put the file on our cluster, I can look at it there.

**Comment by lizziel** [comment]
@tclune, right, my build is parallel so the error message does not necessarily point to pflogger. I did not hit an upload limit - the verbose option when sending stdout and stderr to log file prints the error message to screen instead. Is there a different data stream being used? For the non-verbose build it was short enough to copy to file. For the verbose, not so much.

**Comment by tclune** [comment]
@lizziel   I don't know where the modified flags are coming from in our example.    Our Intel.cmake  has
```
... -check bounds -check uninit ...
```
But your reproducer has
```
... -check bounds uninit ...
```
So some layer is modifying the flags and getting it wrong.   This does not happen with GEOS so I suspect the modifications are happening on your end.    Happy to accept a PR on our end that resolves this, so long as it does not break GEOS.

**Comment by lizziel** [comment]
Interesting. I'll see if I can figure out where that is coming from. @LiamBindle, any ideas?

**Comment by LiamBindle** [comment]
@lizziel @tclune Sorry I didn't see this until now. Thanks for @ me Lizzie.

The problem is the compiler option de-duplicator. CMake runs a step where it removes duplicate flags to avoid passing the same compiler option many times. This is avoided in GEOS because GEOS-ESM/ESMA_cmake sets compiler options like
```cmake
set (CMAKE_Fortran_FLAGS_DEBUG   "${GEOS_Fortran_FLAGS_DEBUG}" CACHE STRING "Debug Fortran flags" FORCE )
```
and the de-duplicator does not run on `CMAKE_Fortran_FLAGS_<CONFIG>`. The de-duplicator runs on options specified via target properties. 

In GCHP we set compiler options like so
```
separate_arguments(GEOS_Fortran_FLAGS_<CONFIG> UNIX_COMMAND ${GEOS_Fortran_FLAGS_RELEASE})
```
and then add `${GEOS_Fortran_FLAGS_<CONFIG>}` to targets via the `target_compile_options()` command. In GCHP we do this to de-couple the compiler arguments of MAPL and GEOS-Chem since `CMAKE_Fortran_FLAGS_<CONFIG>` is a (directory-down) global variable.

---

@lizziel I think this will be fixed by wrapping multiword arguments with `"SHELL:word1 word2 [...]"`. For example, `-check bounds` should be `"SHELL:-check bounds"` and `-check uninit` should be `"SHELL:-check uninit"`. This tells the de-duplicator that `-check bounds` and `-check uninit` are arguments, not `-check`, `bounds`, `-check`, and `uninit` which results in the second `-check` getting dropped.

---

This is actually the same thing that's blocking #57. The holdup for setting comiler options via target properties is that compiler options have to have a consistent ordering in GEOS (for bitwise zero-diff requirement). So for the time being this is GCHP specific.

Note: I think we could actually get consistent option ordering in GEOS via target properties by wrapping the entire contents of `CMAKE_Fortran_FLAGS_<CONFIG>` with `SHELL:`. That should cause the de-duplicator to treat the entire string as one argument, and preserve its current ordering. Perhaps that's something I could explore further.

**Comment by mathomp4** [comment]
@LiamBindle I am looking at changing the Intel Debug flags to just use `-check all`. As long as GEOS runs with it, it's probably a better option anyway!

**Comment by lizziel** [comment]
That would be great. I will try it and see if it works on our end.

**Comment by mathomp4** [comment]
Well the answer is _nope_, at least for `binarytile.x`. Died right out of the gate!
```
forrtl: warning (406): fort: (1): In call to I/O Read routine, an array temporary was created for argument #5

Image              PC                Routine            Line        Source
binarytile.x       000000000040B710  Unknown               Unknown  Unknown
binarytile.x       00000000004051E5  MAIN__                     55  binarytile.F90
binarytile.x       0000000000402A6E  Unknown               Unknown  Unknown
libc-2.22.so       00002AAAAB40D725  __libc_start_main     Unknown  Unknown
binarytile.x       0000000000402969  Unknown               Unknown  Unknown
```
also:
```
forrtl: warning (406): fort: (1): In call to I/O Read routine, an array temporary was created for argument #3
```

On the plus side, at least I got the Intel traceback.

**Comment by tclune** [comment]
@mathomp4 was there any actual error messages?   The warnings are harmless.

**Comment by lizziel** [comment]
That's too bad since it does indeed fix the issue on our end. I'm now running to see what kind of errors get detected with the catchall. The drawback is it is now super slow.

**Comment by tclune** [comment]
@lizziel Don't give up just yet.   We _might_ be able to get this to work with a reasonable investment of effort.

**Comment by mathomp4** [comment]
> @mathomp4 was there any actual error messages? The warnings are harmless.

Nope. In fact, it might be "fun" to debug. I just tried running `binarytile.x` on the command line and it got stuck in essentially a:
```
10 print warning
20 print traceback
30 goto 10
```
loop and pegged a processor. I killed it after about 5-10 seconds and in that time it went through about 5000 iterations of the traceback print. Hmm. 

So, I took a "pre-made" `tile.bin` from another run and moved it over to avoid this and then I get an actual error:
```
forrtl: error (63): output conversion error, unit -5, file Internal Formatted Write
Image              PC                Routine            Line        Source
GEOSgcm.x          00000000117482AE  Unknown               Unknown  Unknown
GEOSgcm.x          00000000117A0AE2  Unknown               Unknown  Unknown
GEOSgcm.x          000000000B3C0FB5  fvdycorecubed_gri        1596  DynCore_GridCompMod.F90
libesmf_fullylink  00002AAACA1C57CB  _ZN5ESMCI6FTable1     Unknown  Unknown
libesmf_fullylink  00002AAACA1C92BF  ESMCI_FTableCallE     Unknown  Unknown
libesmf_fullylink  00002AAACA6A6195  _ZN5ESMCI2VM5ente     Unknown  Unknown
libesmf_fullylink  00002AAACA1C6E7A  c_esmc_ftablecall     Unknown  Unknown
libesmf_fullylink  00002AAACA1C6A2E  _ZN5ESMCI6FTable1     Unknown  Unknown
libesmf_fullylink  00002AAACA1C6765  c_esmc_setservice     Unknown  Unknown
libesmf_fullylink  00002AAACAAC5C44  esmf_gridcompmod_     Unknown  Unknown
GEOSgcm.x          00000000100753E9  mapl_genericmod_m        4560  MAPL_Generic.F90
GEOSgcm.x          000000001007647F  mapl_genericmod_m        4595  MAPL_Generic.F90
GEOSgcm.x          00000000005D28FE  geos_superdyngrid         149  GEOS_SuperdynGridComp.F90
```
but that doesn't seem to stop the model. It goes on and starts throwing tracebacks with every warning! Some of which are:
```
$ rg forrtl 1day.avoid_binarytile.1x6.out  | sort | uniq
forrtl: error (63): output conversion error, unit -5, file Internal Formatted Write
forrtl: warning (406): fort: (1): In call to CELL_CENTER2, an array temporary was created for argument #1
forrtl: warning (406): fort: (1): In call to CELL_CENTER2, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to CELL_CENTER2, an array temporary was created for argument #3
forrtl: warning (406): fort: (1): In call to CELL_CENTER2, an array temporary was created for argument #4
forrtl: warning (406): fort: (1): In call to CELL_CENTER2, an array temporary was created for argument #5
forrtl: warning (406): fort: (1): In call to GET_SYMMETRY, an array temporary was created for argument #1
forrtl: warning (406): fort: (1): In call to GET_SYMMETRY, an array temporary was created for argument #2
forrtl: warning (406): fort: (1): In call to MPI_GROUP_INCL, an array temporary was created for argument #3
```

I wonder if this is a bug or something in Intel 18? I'll try with Intel 19.1.2 and see what happens.

**Comment by lizziel** [comment]
Yes, the traceback per warning is rather excessive. My run failed fairly early since I did not have the AdvCore_GridCompMod.F90 fix (https://github.com/GEOS-ESM/FVdycoreCubed_GridComp/issues/71) and my log was already at 1G.

@LiamBindle commented on a way to fix this in GCHPctm issue https://github.com/geoschem/GCHPctm/issues/37.

> I believe those temporary array warnings can be suppressed with -check,noarg_temp_created.

**Comment by tclune** [comment]
Yes - that would suppress the warning, but ... might bring us right back to the original problem.

sigh

**Comment by mathomp4** [comment]
Well, for now I will try `-check all,noarg_temp_created`. 

Note I did try setting: `setenv FOR_DISABLE_STACK_TRACE 1` which did indeed suppress the stack traces...but it still spitout an *amazing* number of `forrtl` messages: 1539012 before I hit Ctrl-C. 

And this was using 6 processes.

And it only got through about 1.5 time steps of our model. 

I think we make a lot of temp arrays...

**Comment by lizziel** [comment]
Doing the comma-separated list works for me for both ifort18 and ifort19. I still get a few warnings about temporary arrays, but only about 30. It then dies with this, presumably an entirely separate issue I'm looking into so not really relevant here:
```

stdout:
DYNAMICSRun                                    891

stderr (only one thread shown here:
pe=00002 FAIL at line=00067    GetPointer.H                             <status=41>
pe=00002 FAIL at line=01829    MAPL_Generic.F90                         <needs informative message>
pe=00002 FAIL at line=00454    GCHP_GridCompMod.F90                     <status=1>
pe=00002 FAIL at line=01829    MAPL_Generic.F90                         <needs informative message>
pe=00002 FAIL at line=01087    MAPL_CapGridComp.F90                     <status=1>
pe=00002 FAIL at line=01013    MAPL_CapGridComp.F90                     <status=1>
pe=00002 FAIL at line=00733    MAPL_CapGridComp.F90                     <status=1>
pe=00002 FAIL at line=00863    MAPL_CapGridComp.F90                     <status=1>
pe=00002 FAIL at line=00324    MAPL_Cap.F90                             <status=1>
pe=00002 FAIL at line=00198    MAPL_Cap.F90                             <status=1>
pe=00002 FAIL at line=00157    MAPL_Cap.F90                             <status=1>
pe=00002 FAIL at line=00131    MAPL_Cap.F90                             <status=1>
pe=00002 FAIL at line=00029    GCHPctm.F90                              <status=1>

```

**Comment by mathomp4** [comment]
This should be closed by the release of v3.1.4 (:pie:!) which uses the new `-check all,noarg_temp_created`
